packages/model-viewer/src/three-components/CachingGLTFLoader.ts

```diff
@@ -178,17 +180,51 @@ export class CachingGLTFLoader {
         }
         node.castShadow = true;
         const mesh = node as Mesh;
+        let transparent = false;
         const materials =
             Array.isArray(mesh.material) ? mesh.material : [mesh.material];
         materials.forEach(material => {
           if ((material as any).isMeshStandardMaterial) {
+            if (material.transparent && material.side === DoubleSide) {
+              transparent = true;
+              material.side = FrontSide;
+            }
             this.roughnessMipmapper.generateMipmaps(
                 material as MeshStandardMaterial);
           }
         });
+
+        if (transparent) {
+          duplicate.push(mesh);
+        }
       });
     }
 
+    // We duplicate transparent, double-sided meshes and render the back face
+    // before the front face. This creates perfect triangle sorting for all
+    // convex meshes. Sorting artifacts can still appear when you can see
+    // through more than two layers of a given mesh, but this can usually be
+    // mitigated by the author splitting the mesh into mostly convex regions.
+    // The performance cost is not too great as the same shader is reused and
+    // the same number of fragments are processed; only the vertex shader is run
+    // twice.
+    for (let i = 0; i < duplicate.length; i++) {
+      const mesh = duplicate[i];
+      const material = Array.isArray(mesh.material) ?
```

Comment:
I think it's a bit easier to follow the traversal scheme used above on line 184. There, we wrap `mesh.material` in an array if it is not an array. This is arguably wasteful because material is usually _not_ an array, but at the end of the day it is cheap enough that you won't notice it in a profile, and it has the advantage of simplifying the subsequent traversal code.

Reply:
Unfortunately three does not accept a material array when it doesn't expect one, so I need a little extra tweaking, but I still like the look of this way better than how it was. :+1: